### PR TITLE
FactoryDesignPatternLerning_StepOne

### PR DESCRIPTION
**Pull request by JHamide**
I started doing my homework, but I'm taking it step by step based on yesterday's meeting and gradually completing it.
@mohammadKarimi

### CODE REVIEW

File: Examples/Jalilpour/DeliveryFactory.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        Air,
+        Ship,
+        Train
+    }
+
+    public class DeliveryService
+    {
+        public Delivery Create(CargoType cargoType)
+        {
+            if (cargoType == CargoType.Air)
```
Comment (mohammadKarimi) [inline Examples/Jalilpour/DeliveryFactory.cs:15]: you have created the factory for each type, but why are you making instances by your self?
you have to seperate concern to each factory to make instances. and manage the instances. 
for example : if Air must be single instance the factory of Air is responsible to do it.

File: Examples/Jalilpour/DeliveryFactory.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            else if (cargoType == CargoType.Train)
+                return new Train();
+            else if (cargoType == CargoType.Ship)
+                return new Ship();
+            else
+                throw new ArgumentException("Invalid Cargo Type");
+        }
+    }
+
+    public abstract class Delivery
```
Comment (mohammadKarimi) [inline Examples/Jalilpour/DeliveryFactory.cs:26]: it is good.

File: Examples/Jalilpour/DeliveryFactory.cs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+            var instance = deliverFactory.Create(CargoType.Air);
+        }
+    }
+
+    public abstract class DeliveryFactory
+    {
+        public abstract Delivery CreateDelivery();
+    }
+
+    public class AirDeliveryFactory : DeliveryFactory
```
Comment (mohammadKarimi) [inline Examples/Jalilpour/DeliveryFactory.cs:69]: you have to manage the instance of Air here.

**Review by mohammadKarimi — CHANGES_REQUESTED** [review]

File: Examples/Jalilpour/DeliveryFactory.cs [hunk 4]
```diff
@@ -13,13 +13,18 @@ public class DeliveryService
...
-                return new Train();
+            {
+                return new AirDeliveryFactory().CreateDelivery();
+            }
             else if (cargoType == CargoType.Ship)
-                return new Ship();
-            else
-                throw new ArgumentException("Invalid Cargo Type");
+            {
+                return new ShipDeliveryFactory().CreateDelivery();
```
Comment (mohammadKarimi) [inline Examples/Jalilpour/DeliveryFactory.cs:21]: it is better. good job.

**Review by mohammadKarimi — COMMENTED** [review]
now move to harder one.

Imagine it is a library and we make it as a package. so users or customers, can't change your code, but they want to add a new behavior like Truck. so, how you can solve it.

File: Examples/Jalilpour/DeliveryFactory.cs (hunk 1 above)
Comment (JHamide) [reply Examples/Jalilpour/DeliveryFactory.cs:15]: Off Course, I'm going to correct that.

**Comment by JHamide** [comment]
> now move to harder one.
> 
> Imagine it is a library and we make it as a package. so users or customers, can't change your code, but they want to add a new behavior like Truck. so, how you can solve it.

For this question I have to use reflection? I have no idea, Could you please guide me?

File: Examples/Jalilpour/DeliveryFactory.cs
Comment (JHamide) [inline Examples/Jalilpour/DeliveryFactory.cs:1]: @mohammadKarimi  #Request a review ===> Add a new behavior like Truck

**Pull request closed without merging**